lambda_functions/va_profile/va_profile_opt_in_out_lambda.py

```diff
@@ -448,6 +493,157 @@ def make_PUT_request(
         https_connection.close()
 
 
+def send_comp_and_pen_opt_in_confirmation(va_profile_id: int) -> Optional[HTTPResponse]:
+    """
+    Send an opt-in confirmation SMS notification based on user's VAProfile ID.
+
+    Args:
+        va_profile_id (int): The VA profile ID of the user the notification should be sent to.
+    """
+
+    try:
+        # Personalization for opt-in confirmation notification SMS based on cutoff date
+        # to enroll in monthly Comp and Pen notification
+        now = datetime.now(timezone.utc)
+        cutoff_datetime = datetime(now.year, now.month, *COMP_AND_PEN_OPT_IN_CUTOFF_TIME_UTC, tzinfo=timezone.utc)
+
+        if now < cutoff_datetime:
+            month_personalisation = calendar.month_name[now.month]
+        else:
+            next_month = (now.month % 12) + 1
+            month_personalisation = calendar.month_name[next_month]
+
+        sms_data = json.dumps(
+            {
+                'template_id': COMP_AND_PEN_OPT_IN_TEMPLATE_ID,
+                'recipient_identifier': {'id_type': 'VAPROFILEID', 'id_value': str(va_profile_id)},
+                'sms_sender_id': COMP_AND_PEN_SMS_SENDER_ID,
+                'personalisation': {'month-name': month_personalisation},
+            }
+        )
+
+        logger.debug('Sending Comp and Pen opt-in confirmation SMS notification vaProfileId %s', va_profile_id)
+
+        conn = HTTPSConnection('{}.api.notifications.va.gov'.format(NOTIFY_ENVIRONMENT), context=ssl_context)
+        encoded_header = generate_jwt(COMP_AND_PEN_OPT_IN_API_KEY, COMP_AND_PEN_SERVICE_ID)
+
+        conn.request(
+            'POST',
+            '/v2/notifications/sms',
+            body=sms_data,
+            headers={'Content-Type': 'application/json', 'Authorization': 'Bearer {}'.format(encoded_header)},
+        )
+
+        response = conn.getresponse()
+
+        if response.status == 201:
+            logger.info(
+                'Comp and Pen opt-in confirmation SMS notification sent successfully. Response status: %s',
```

Comment:
Should we log the notification_id since we got a 201?

Reply:
So this is dumb and I spent about an hour on this. I ran into an issue while trying to log the notification_id from the HTTPS response. The core of the problem is related to how the response object's data is handled when using the HTTPSConnection library.

Specifically, the response.read().decode() method reads the data stream of the response, but it also consumes it. This means that once the data has been read, it’s no longer available for subsequent calls—essentially, the stream gets exhausted.

So, if I use response.read().decode() to get the notification_id early on, the response object becomes empty, and I can’t access the data again later in the code. That's why I can't both log the notification_id right away and still have the complete response available when returning from the Lambda handler.


I have updated the code to log the notification_id in the lambda handler after this method returns this response  object.

Reply:
I hear you, which is why Requests could be used (and I understand why you didn't).